rfcs/hse/0008-rework-cn-stats.md

```diff
@@ -0,0 +1,182 @@
+# Rework Internal CN Tree Statistics
+
+## Problem
+
+Internal CN tree statistics are confusing, redundant, misleading and inefficient.
+
+1. There is much duplication caused by the same statistic appearing in different
+   structures.  For example, `key_stats.tot_vlen`, `kblk_metrics.tot_val_bytes`,
+   and `kvset_metrics.tot_val_bytes` all track the same statistic.
+1. `kvset_metric` and `kvset_stats` are redundant. We should pick one.
+1. Many metrics don't measure what their name implies.  See *Misleading statistics*.
+1. `kvset_metrics` is computed on demand, requires a node lock, and has to
+   iterate over all kblocks in the kvset -- despite the fact that they never
+   change for the life of a kvset.
+1. Some kvset level statistics are stored in CNDB, others are stored in kblock
+   headers, and others are stored in the kvset hblock.
+
+### Misleading statistics
+
+1. `key_stats.nvals` measures number of values associated with key.  It includes
+   tombstones but not prefix tombstones.
+1. `key_stats.tot_vlen`, `kblk_metrics.tot_val_bytes`,
+   `kvset_metrics.tot_val_bytes`: These measure the total on-media length of all
+   values associated with key/kblock/kvset.  It would be nice to have more
+   detail such as: length of vals in kblocks (VTYPE_IVAL) or maybe just lengths
+   of values by type (enum kmd_type).
+1. `kvset_metrics.num_keys` measures number of keys. It includes keys with
+   tombstones but not keys with prefix tombstones.
```

Comment:
`tombstones, but`

Reply:
Comma only needed when connecting two independent clauses, which is not the case here.